app/src/org/commcare/tasks/DeleteLogs.kt

```diff
@@ -0,0 +1,59 @@
+package org.commcare.tasks
+
+import android.content.Context
+import androidx.work.Worker
+import androidx.work.WorkerParameters
+import org.commcare.CommCareApplication
+import org.commcare.android.javarosa.AndroidLogEntry
+import org.commcare.android.javarosa.DeviceReportRecord
+import org.commcare.android.logging.ForceCloseLogEntry
+import org.commcare.logging.XPathErrorEntry
+import org.commcare.models.database.SqlStorage
+import org.commcare.preferences.HiddenPreferences
+import org.javarosa.core.model.utils.DateUtils
+import java.io.File
+import java.util.Date
+import java.util.Vector
+
+
+// A Worker class used for deleting any logs older than 3 months
+class DeleteLogs(appContext: Context, workerParams: WorkerParameters)
+    : Worker(appContext, workerParams) {
+
+    override fun doWork(): Result {
+        val today = Date().time
+        val threeMonthsAgo = Date(today - (90L * DateUtils.DAY_IN_MS));
+
+        // Clear logs older than 3 month stored in DB
+        purge(CommCareApplication.instance().getUserStorage(AndroidLogEntry.STORAGE_KEY, AndroidLogEntry::class.java), threeMonthsAgo)
+        purge(CommCareApplication.instance().getUserStorage(XPathErrorEntry.STORAGE_KEY, XPathErrorEntry::class.java), threeMonthsAgo)
+        purge(CommCareApplication.instance().getGlobalStorage(ForceCloseLogEntry.STORAGE_KEY, ForceCloseLogEntry::class.java), threeMonthsAgo)
+        purge(CommCareApplication.instance().getUserStorage(XPathErrorEntry.STORAGE_KEY, XPathErrorEntry::class.java), threeMonthsAgo)
+
+        // Clear logs older than 3 month which have been serialized to disk
+        val deviceReportRecordStorage = CommCareApplication.instance().getUserStorage(DeviceReportRecord::class.java)
+        val reportsToDelete = Vector<Int>()
+        for (deviceReportRecord in deviceReportRecordStorage) {
+            val reportFile = File(deviceReportRecord.filePath)
+            if (reportFile.lastModified() < threeMonthsAgo.time) {
```

Comment:
In the event that this code completes partially (IE: we make it 1/2 through the Storage loop, but never make it to line 43 below), there will be old historical DeviceReportRecords with references to Files that no longer exist. 

This line would need to include some provision for the file not existing to protect against that, but more generally there may be other uses of the DeviceReportRecord which will also need to be updated in the same way. 

I'm not sure how the records are cleared out in other contexts, since presumably they have the same transactional atomicity.

Reply:
@ctsims  The `lastModified` returns 0 in case the file doesn't exist and the `delete` method doesn't throw an exception if file doesn't exist. So I think it's safe to leave it as it is. Verified the same at other places as well.